src/solvers/fd_tree/fd_2d.jl

```diff
@@ -0,0 +1,199 @@
+      mul!(view(du_vectors, :, j, element), D, view(f_element, :, j),
+           one(eltype(du)), one(eltype(du)))
+    end
+
+    # TODO: FD. Doesn't work until https://github.com/JuliaArrays/ArrayInterface.jl/pull/156
+    #           is merged and the other issue mentioned there is fixed.
+    # y direction
+    @. f_element = flux(u_element, 2, equations)
+    for i in eachnode(dg)
+      mul!(view(du_vectors, i, :, element), D, view(f_element, i, :),
+           one(eltype(du)), one(eltype(du)))
+    end
+  end
+
+  return nothing
+end
+
+
+function prolong2mortars!(cache, u, mesh, equations, mortar::Nothing,
+                          surface_integral, dg::DG)
+  @assert isempty(eachmortar(dg, cache))
+end
+
+function calc_mortar_flux!(surface_flux_values, mesh,
+                           nonconservative_terms, equations,
+                           mortar::Nothing,
+                           surface_integral, dg::DG, cache)
+  @assert isempty(eachmortar(dg, cache))
+end
+
+
+function calc_surface_integral!(du, u, mesh::TreeMesh{2},
+                                equations, surface_integral::SurfaceIntegralStrongForm,
+                                dg::DG, cache)
+  inv_weight_left  = inv(left_boundary_weight(dg.basis))
+  inv_weight_right = inv(right_boundary_weight(dg.basis))
+  @unpack surface_flux_values = cache.elements
+
+  @threaded for element in eachelement(dg, cache)
+    for l in eachnode(dg)
+      # surface at -x
+      u_node = get_node_vars(u, equations, dg, 1, l, element)
+      f_node = flux(u_node, 1, equations)
+      f_num  = get_node_vars(surface_flux_values, equations, dg, l, 1, element)
+      surface_contribution = -(f_num - f_node) * inv_weight_left
+      add_to_node_vars!(du, surface_contribution, equations, dg, 1, l, element)
+
+      # surface at +x
+      u_node = get_node_vars(u, equations, dg, nnodes(dg), l, element)
+      f_node = flux(u_node, 1, equations)
+      f_num  = get_node_vars(surface_flux_values, equations, dg, l, 2, element)
+      surface_contribution = +(f_num - f_node) * inv_weight_right
+      add_to_node_vars!(du, surface_contribution, equations, dg, nnodes(dg), l, element)
+
+      # surface at -y
+      u_node = get_node_vars(u, equations, dg, l, 1, element)
+      f_node = flux(u_node, 2, equations)
+      f_num  = get_node_vars(surface_flux_values, equations, dg, l, 3, element)
+      surface_contribution = -(f_num - f_node) * inv_weight_left
+      add_to_node_vars!(du, surface_contribution, equations, dg, l, 1, element)
+
+      # surface at +y
+      u_node = get_node_vars(u, equations, dg, l, nnodes(dg), element)
+      f_node = flux(u_node, 2, equations)
+      f_num  = get_node_vars(surface_flux_values, equations, dg, l, 4, element)
+      surface_contribution = +(f_num - f_node) * inv_weight_right
+      add_to_node_vars!(du, surface_contribution, equations, dg, l, nnodes(dg), element)
+    end
+  end
+
+  return nothing
+end
+
+
+# AnalysisCallback
```

Comment:
So this would eventually move into a file like `analysis_fd2d.jl` right?

Reply:
Exactly, that would be part of the clean-up :+1:

Reply:
(tracked in #607)